### DISCUSSION TITLE
How to Setup a Unified Processor in Typescript With Rehype?

### DISCUSSION THREAD

**Discussion by manunamz** [remarkjs/remark#1031] [Q&A] [answered]
Hello,

I'm looking at a remark plugin similar in format to [remark-cite](https://github.com/benrbray/remark-cite) (mono-repo, typescript, etc.). I want to write some unit tests to verify the shape of `mdast` data and ensure it will work with [`rehype`](https://github.com/rehypejs/rehype), as well as provide functional documentation on how to setup the `remark` plugin with `rehype`. 

In order to setup a unified `processor`, I have referenced:

- [unified overview](https://github.com/unifiedjs/unified#overview).  
- [rehype-parse](https://github.com/rehypejs/rehype/tree/main/packages/rehype-parse#use)...
  - ...for my `html -> markdown` tests and 
- [rehype-stringify](https://github.com/rehypejs/rehype/tree/main/packages/rehype-stringify#use)... 
  - ...for my `markdown -> html` tests.

But I am getting type errors when I have a test executor like so (for the `html -> markdown` case, though both are effected):

```ts
import * as assert from 'assert';
import unified from 'unified';
import remarkStringify from 'remark-stringify';
import rehypeRemark from 'rehype-remark';
import rehypeParse from 'rehype-parse';
import { remarkPlugin } from '../src';

const testRemarkHtmlToMarkdown = (params: any) => function () { 
  // setup vars
  var opts: Partial<PluginOptions> = params.opts;
  var html: string = params.html;
  var expdMkdn: string = params.mkdn;
  // init
  const processor = unified()
                        .use(rehypeParse)   // <--- error
                        .use(rehypeRemark)  // <--- error
                        .use(remarkStringify)
                        .use(remarkPlugin, opts);
  // go
  var actlMkdn: string = String(processor.processSync(html));
  // test
  assert.strictEqual(actlMkdn, expdMkdn);
};
```

The `error` is:

```
No overload matches this call.
  The last overload gave the following error.
    Argument of type 'Plugin<[(Options | undefined)?] | void[], string, Root>' is not assignable to parameter of type 'ProcessorSettings<Settings>'.ts(2769)
index.d.ts(55, 5): The last overload is declared here.
```

I tried fixing the errors with type casting:

```ts
import unified, { Plugin } from 'unified';
  // ...
  const processor = unified()
                        .use(rehypeParse as unknown as Plugin)   // <--- type casted
                        .use(rehypeRemark as unknown as Plugin)  // <--- type casted
                        .use(remarkStringify)
                        .use(remarkWikiLinksPlugin, opts);
  // ...
};
```

But re-running the tests yields further errors. I will refrain from posting the actual errors, as these particular ones have proved inaccurate in the past...But it does hint that the type casting still isn't working...

All type complaints are confined to `rehype`. I have in my package dependencies:

- `@types/unist`
- `@types/mdast`
- `@types/hast`

How can I setup the unified `processor` in typescript, as shown in [rehype-parse](https://github.com/rehypejs/rehype/tree/main/packages/rehype-parse#use) and [rehype-stringify](https://github.com/rehypejs/rehype/tree/main/packages/rehype-stringify#use)?

Thanks much, unified is really a nice project to work with. 🙏

PS: I'm not 100% sure these tests are the best use of code-space, but a part of this exercise is to prove to myself that I understand how `remark` and `rehype` will fit together -- especially since I'm not really working with `rehype`. So, advice on this test approach is also welcome.

**Comment by JounQin** [comment]
A reproduction would help much.

**Comment by ChristianMurphy** [comment, marked as answer]
@manunamz is your question about `remark` or `rehype`?
For rehype it may be better to ask at https://github.com/rehypejs/rehype/discussions for clarity and ease of finding it.

For remark, unified and remark are compatible, my guess is you may have incompatible major versions of dependencies.
As @JounQin notes a minimal sandbox goes a long way to better understanding what you are asking, some starters could include:


| description | codesandbox | stackblitz |
| - | - | - |
| remark only (for markdown to markdown) | [codesandbox](https://codesandbox.io/s/github/remarkjs/.github/tree/main/sandbox-templates/remark-with-parcel) | [stackblitz](https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/remark-with-parcel) |
| remark and rehype (for markdown to html) | [codesandbox](https://codesandbox.io/s/github/remarkjs/.github/tree/main/sandbox-templates/remark-rehype-with-parcel) | [stackblitz](https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/remark-rehype-with-parcel) |
| react-markdown | [codesandbox](https://codesandbox.io/s/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-create-react-app) | [stackblitz](https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-create-react-app) |

**Reply by ChristianMurphy** [reply]
Confirmed the major version being incompatible was the issue https://github.com/orgs/remarkjs/discussions/1031#discussioncomment-3425926

**Reply by manunamz** [reply]
I left this issue for a while because I was really reluctant to upgrade. `remark-rehype` `v9+` is [`esm-only`](https://github.com/remarkjs/remark-rehype/releases/tag/9.0.0). 

Originally I wanted to do an initial release that was both `commonjs`/`esm`-friendly, so if anyone was stuck on an older version of the `unified` project, there would be _a_ version of my plugins they could use.

But upgrading is in fact how to get past this issue.

**Comment by wooorm** [comment]
Your `remarkPlugin` is most likely not typed, or typed incorrectly.
Look at existing plugins to see how to type them.
Or share your plugin as well for help.

**Comment by manunamz** [comment]
Thanks for the sandbox link and the plugin type advice -- the problem does seem to be in that direction.

An in-context example has been committed [here](https://github.com/benrbray/remark-cite/commit/684d54588f1012244f208a610f5e81904edb1c40). You can clone the repo and play with the full state on the [`test_markhype`](https://github.com/manunamz/remark-cite/tree/test_markhype) branch at the linked commit, cd into `remark-cite` subproject, run `npm install`, then `npm run test` to see the output. Output should look like [this](https://gist.github.com/manunamz/93834209fdd394c3d9eadf461a341f8d#import-method-2) depending on which import method you've used.

There is also a sandbox [here](https://codesandbox.io/s/summer-cache-x4bl9q?file=/src/index.ts:969-1021), which exhibits a similar, but not the same, typescript error[^error]. This is helpful insofar as it illustrates the static error, but unhelpful as it, curiously, runs despite the error (though the behavior across browser tabs has not been consistent).

🙏 Thanks again, for the quick help and attention. And apologies for the silly error-ridden question which is probably going to end up being solved by toggling a setting somewhere...

(question from title remains)

[^error]: Error from the sandbox is below (contrasted to the error when running locally above):
    ```
    No overload matches this call.
      The last overload gave the following error.
        Argument of type 'Plugin<[] | [any, Options?] | [Options], Root>' is not assignable to parameter of type 'ProcessorSettings<Settings>'.
          Property 'settings' is missing in type 'Plugin<[] | [any, Options?] | [Options], Root>' but required in type 'ProcessorSettings<Settings>'.ts(2769)
    ```

**Reply by wooorm** [reply]
`remark-city` isn’t typed with the required generics. This can most likely be solved there. Raise an issue there.

**Reply by wooorm** [reply]
The dependencies are out of date in your sandbox. Update them

**Reply by ChristianMurphy** [reply]
The answer is in https://github.com/orgs/remarkjs/discussions/1031#discussioncomment-3419798
Taking your sandbox, and updating `remark-rehype` to the next major version which is compatible with unified's typings (`remark-rehype` version 9 was not version 10 is) solves the issue https://codesandbox.io/s/throbbing-wind-x6nhi8?file=/src/index.ts
